### ISSUE TITLE
Feature Request: Create kaira-train Training Console Script

### ISSUE THREAD

**Issue by selimfirat** [ipc-lab/kaira#40]
# Feature Request: Create kaira-train Training Console Script

## Summary
Create a robust, user-friendly training console script for the Kaira framework, similar to the existing `kaira-benchmark` command, that provides comprehensive command-line interface for training communication system models.

## Motivation
Currently, users need to write custom Python scripts to train models in Kaira. A dedicated CLI tool would:
- Simplify the training workflow for researchers and practitioners
- Provide standardized parameter handling across different models
- Enable batch training and experimentation
- Integrate with existing Kaira infrastructure (ModelRegistry, etc.)
- Support both beginners and advanced users with flexible configuration options

### Core Functionality
- **Model Training**: Support for all registered communication models
- **CLI Interface**: Clean command-line interface with comprehensive help
- **Configuration Support**: Multiple configuration methods (CLI args, YAML, JSON)
- **Model Discovery**: Ability to list available models
- **Output Management**: Structured output directories with checkpoints and logs

### Command-Line Interface
```bash
# Basic usage
kaira-train --model deepjscc --output-dir ./results

# List available models
kaira-train --list-models

# Configuration file support
kaira-train --model deepjscc --config-file ./configs/training.yaml

# Communication-specific parameters
kaira-train --model channel_code \
  --snr-min 0 --snr-max 15 \
  --channel-uses 128 \
  --channel-type awgn
```

### Configuration Support
1. **Command-line arguments**: Direct parameter specification
2. **Hydra YAML**: Advanced configuration with interpolation and composition
3. **JSON**: Simple structured configuration
4. **Mixed mode**: CLI args override config file parameters

### Training Parameters
- **Basic**: epochs, batch size, learning rate, warmup steps
- **Communication-specific**: SNR ranges, channel types, noise parameters
- **Performance**: device selection, mixed precision, data loading workers
- **Monitoring**: evaluation strategies, logging, checkpointing

### Integration Requirements
- **ModelRegistry**: Use existing model registration system
- **Hydra**: Support for advanced configuration management
- **OmegaConf**: Configuration parsing and validation
- **Transformers**: Leverage HuggingFace training infrastructure where applicable

### File Structure
```
scripts/
├── kaira_train.py          # Main CLI script
└── __init__.py

configs/
├── training_example.yaml   # Example Hydra configuration
└── training_simple.yaml   # Simple configuration template

setup.py                   # Add entry point for kaira-train
```

### Entry Point
Add to `setup.py`:
```python
entry_points={
    "console_scripts": [
        "kaira-benchmark=scripts.kaira_benchmark:main",
        "kaira-train=scripts.kaira_train:main",  # New entry point
    ],
},
```

### Configuration Examples

**Hydra YAML** (`configs/training_example.yaml`):
```yaml
# @package _global_

defaults:
  - _self_

model:
  _target_: kaira.models.DeepJSCCModel
  type: deepjscc
  input_dim: 512
  channel_uses: 64

training:
  output_dir: ./training_results
  num_train_epochs: 10
  per_device_train_batch_size: 32
  learning_rate: 1e-4
  snr_min: 0.0
  snr_max: 20.0
  channel_type: awgn

hydra:
  run:
    dir: ${training.output_dir}/hydra_outputs/${now:%Y-%m-%d_%H-%M-%S}
```

**JSON** (`configs/training_simple.json`):
```json
{
  "model": {
    "type": "deepjscc",
    "input_dim": 512,
    "channel_uses": 64
  },
  "training": {
    "output_dir": "./training_results",
    "num_train_epochs": 10,
    "per_device_train_batch_size": 32,
    "learning_rate": 1e-4,
    "snr_min": 0.0,
    "snr_max": 20.0
  }
}
```

### Documentation Page
Create comprehensive documentation similar to the benchmark docs:
- Installation and setup instructions
- Quick start guide with examples
- Complete parameter reference
- Configuration file documentation
- Model integration guide
- Troubleshooting section

### Integration
- Add to main documentation index under "Getting Started"
- Cross-reference with API documentation
- Include in examples and tutorials

### Functional Requirements
- [ ] CLI script executable as `kaira-train`
- [ ] Support for all registered models via ModelRegistry
- [ ] Command-line argument parsing with comprehensive help
- [ ] YAML configuration file support (Hydra format)
- [ ] JSON configuration file support
- [ ] Model listing functionality (`--list-models`)
- [ ] Structured output directory with checkpoints and logs
- [ ] Integration with existing training infrastructure

### Quality Requirements
- [ ] Comprehensive error handling and user-friendly error messages
- [ ] Input validation for all parameters
- [ ] Proper logging and progress reporting
- [ ] Memory-efficient training with large models
- [ ] Cross-platform compatibility (Linux, macOS, Windows)

### Documentation Requirements
- [ ] Complete RST documentation page
- [ ] Integration into main documentation index
- [ ] Example configuration files
- [ ] CLI help documentation
- [ ] Troubleshooting guide

### Testing Requirements
- [ ] Unit tests for argument parsing
- [ ] Integration tests with sample models
- [ ] Configuration file loading tests
- [ ] Error handling tests
- [ ] End-to-end training workflow tests

## Dependencies
- `omegaconf`: Configuration management
- `hydra-core`: Advanced configuration features
- `transformers`: Training infrastructure (if applicable)
- Existing Kaira dependencies

## Related Issues
- Similar to existing `kaira-benchmark` script architecture
- Should integrate with current ModelRegistry system
- May require updates to model base classes for training standardization

## Additional Considerations
- **Backward Compatibility**: Ensure existing training workflows continue to work
- **Performance**: Optimize for both small experiments and large-scale training
- **Extensibility**: Design for easy addition of new parameters and features
- **Community**: Consider common use cases in communication systems research
- **Integration**: Potential future integration with experiment tracking tools (W&B, MLflow)

### Research Workflow
```bash
# Quick experiment
kaira-train --model deepjscc --epochs 5 --batch-size 16

# Full experiment with custom config
kaira-train --model channel_code --config-file ./experiments/ldpc_training.yaml

# Hyperparameter sweep preparation
kaira-train --model deepjscc --config-file ./sweep/config_1.yaml --output-dir ./results/run_1
```

### Production Training
```bash
# Large-scale training with monitoring
kaira-train --model production_jscc \
  --config-file ./production/config.yaml \
  --fp16 \
  --dataloader-num-workers 8 \
  --output-dir ./production_results
```

This feature would significantly improve the user experience and make Kaira more accessible to researchers and practitioners in the communication systems field.

**Comment by selimfirat** [comment]
Marking this as resolved.
